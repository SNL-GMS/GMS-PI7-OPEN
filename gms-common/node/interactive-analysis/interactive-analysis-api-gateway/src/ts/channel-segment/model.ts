/**
 * Channel Segment definitions shared across gateway data APIs
 */
import { CreationInfo, Location, OSDCreationInfo, ProcessingContext } from '../common/model';
import { Waveform, OSDWaveform } from '../waveform/model';
import { FkPowerSpectra, FkPowerSpectraOSD, PhaseType } from './model-spectra';

/**
 * Represents a channel time segment comprised of one or more waveforms for a (derived or raw).
 */
export interface ChannelSegment<T extends TimeSeries> {
    id: string;
    channelId: string;
    name: string;
    type: ChannelSegmentType;
    timeseriesType: TimeSeriesType;
    startTime: number;
    endTime: number;
    timeseries: T[];
    creationInfo: CreationInfo;
}

/**
 * OSD channel segment as a generic of OSD time series
 */
export interface OSDChannelSegment<T extends OSDTimeSeries> {
    id: string;
    channelId: string;
    name: string;
    type: ChannelSegmentType;
    timeseriesType: TimeSeriesType;
    startTime: string;
    endTime: string;
    timeseries: T[];
    creationInfo: OSDCreationInfo;
}

/**
 * Channel segment input
 */
export interface ChannelSegmentInput {
    'start-time': string;
    'end-time': string;
    'channel-ids': string[];
    'with-waveforms': boolean;
}

/**
 * Channel segment by id input
 */
export interface ChannelSegmentByIdInput {
    ids: string[];
    'with-waveforms': boolean;
}

/**
 * Represents a generic time series data set output from a channel
 * (derived or raw) within a specified time range.
 */
export interface TimeSeries {
    startTime: number;
    sampleRate: number;
    sampleCount: number;
}

/**
 * Time series OSD representation
 */
export interface OSDTimeSeries {
    startTime: string;
    sampleRate: number;
    sampleCount: number;
}

/**
 * Channel Segment type
 */
export enum ChannelSegmentType {
    ACQUIRED = 'ACQUIRED',
    RAW = 'RAW',
    DETECTION_BEAM = 'DETECTION_BEAM',
    FK_BEAM = 'FK_BEAM',
    FK_SPECTRA = 'FK_SPECTRA',
    FILTER = 'FILTER',
}

/**
 * Time series type
 */
export enum TimeSeriesType {
    WAVEFORM = 'WAVEFORM',
    FK_SPECTRA = 'FK_SPECTRA',
    DETECTION_FEATURE_MAP = 'DETECTION_FEATURE_MAP'
}

/**
 * Beam Input Definitions for Compute Beam Streaming Call
 */
export interface BeamFormingInput {
    beamDefinition: BeamDefinition;
    processingContext: ProcessingContext;
    outputChannelId: string;
    waveforms: OSDChannelSegment<OSDTimeSeries>[];
}

/**
 * Beam Definition used by BeamFormingInput
 */
export interface BeamDefinition {
    phaseType: PhaseType;
    azimuth: number;
    slowness: number;
    coherent: boolean;
    snappedSampling: boolean;
    twoDimensional: boolean;
    nominalWaveformSampleRate: number;
    waveformSampleRateTolerance: number;
    beamPoint: Location;
    // OSD "map" is not really a map so type as any
    relativePositionsByChannelId: any;
    minimumWaveformsForBeam: number;
}

/**
 * Relative Positions By Channel Id Definition
 */
export interface RelativePosition {
    northDisplacementKm: number;
    eastDisplacementKm: number;
    verticalDisplacementKm: number;
}

/**
 * Checks if FK spectra channel segment
 * @param object Channel Segment
 * @returns boolean
 */
// tslint:disable-next-line:max-line-length
export function isFkSpectraChannelSegment(object: ChannelSegment<TimeSeries>): object is ChannelSegment<FkPowerSpectra> {
    return object.timeseriesType === TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks id FK spectra channel segment OSD representation
 * @param object OSDChannelSegment
 * @returns boolean
 */
// tslint:disable-next-line:max-line-length
export function isFkSpectraChannelSegmentOSD(object: OSDChannelSegment<OSDTimeSeries>): object is OSDChannelSegment<FkPowerSpectraOSD> {
    return object.timeseriesType === TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks if FK Spectra time series
 * @param object Time Series
 * @param type Time Series Type
 * @returns boolean
 */
export function isFkSpectraTimeSeries(object: TimeSeries, type: TimeSeriesType): object is FkPowerSpectra {
    return type === TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks if FK Spectra time series
 * @param object Time Series
 * @param type Time Series Type
 * @returns boolean
 */
export function isWaveformTimeSeries(object: TimeSeries, type: TimeSeriesType): object is Waveform {
    return type === TimeSeriesType.WAVEFORM;
}

/**
 * Checks if waveform channel segment
 * @param object Channel segment
 * @returns boolean
 */
export function isWaveformChannelSegment(object: ChannelSegment<TimeSeries>): object is ChannelSegment<Waveform> {
    return object.timeseriesType === TimeSeriesType.WAVEFORM;
}

/**
 * Checks if waveform channel segment OSD representation
 * @param object Channel segment OSD representation
 * @returns boolean
 */
export function isOSDWaveformChannelSegment(
    object: OSDChannelSegment<OSDTimeSeries>): object is OSDChannelSegment<OSDWaveform> {
    return object.timeseriesType === TimeSeriesType.WAVEFORM;
}

/**
 * Checks if channel segment OSD representation
 * @param object Channel segment or channel segment OSD representation
 * @returns boolean
 */
export function isOSDChannelSegment(
    object: ChannelSegment<TimeSeries> | OSDChannelSegment<OSDTimeSeries>): object is OSDChannelSegment<OSDTimeSeries> {
    return typeof object.startTime === 'string';
}

/**
 * Checks if channel segment
 * @param object Channel segment or channel segment OSD representation
 * @returns boolean
 */
export function isChannelSegment(
    object: ChannelSegment<TimeSeries> | OSDChannelSegment<OSDTimeSeries>): object is ChannelSegment<TimeSeries> {
    return typeof object.startTime === 'number';
}

/**
 * Checks if time series OSD representation
 * @param object Time series or Time Series OSD representation
 */
export function isOSDTimeSeries(object: TimeSeries | OSDTimeSeries): object is OSDTimeSeries {
    return typeof object.startTime === 'string';
}

/**
 * Checks if time series
 * @param object Time series or time series OSD representation
 * @returns boolean
 */
export function isTimeSeries(object: TimeSeries | OSDTimeSeries): object is TimeSeries {
    return typeof object.startTime === 'number';
}
