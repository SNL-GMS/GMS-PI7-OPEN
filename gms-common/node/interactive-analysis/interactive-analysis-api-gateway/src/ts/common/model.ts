import { SignalDetection } from '../signal-detection/model';
import { Event } from '../event/model';

/**
 * Common model definitions shared across gateway data APIs
 */

/**
 * Represents a location specified using latitude (degrees), longitude (degrees),
 * and altitude (kilometers).
 */
export interface Location {
    latDegrees: number;
    lonDegrees: number;
    elevationKm: number;
    depthKm?: number;
}

/**
 * Represents a frequency range
 */
export interface FrequencyBand {
    minFrequencyHz: number;
    maxFrequencyHz: number;
}

/**
 * Time range
 */
export interface TimeRange {
    startTime: number;
    endTime: number;
}

/**
 * Represents events and sds to publish
 */
export interface AssociationChange {
    events: Event[];
    sds: SignalDetection[];
}

/**
 * Used when we need to return the original detections mutations where performed on
 * in addition to the other data changed
 */
export interface DetectionAndAssociationChange {
    detections: SignalDetection[];
    associationChange: AssociationChange;
}

/**
 * Used when we need to return the original event mutated where performed on
 * in addition to the other data changed
 */
export interface EventAndAssociationChange {
    event: Event;
    associationChange: AssociationChange;
}

/**
 * Enumeration representing the different types of stations in the monitoring network.
 */
export enum StationType {
    Seismic3Component = 'Seismic3Component',
    Seismic1Component = 'Seismic1Component',
    SeismicArray = 'SeismicArray',
    Hydroacoustic = 'Hydroacoustic',
    HydroacousticArray = 'HydroacousticArray',
    Infrasound = 'Infrasound',
    InfrasoundArray = 'InfrasoundArray',
    Weather = 'Weather',
    UNKNOWN = 'UNKNOWN'
}

/**
 * Enumerated set of channel types
 */
export enum ChannelType {
    BroadbandHighGainVertical = 'BroadbandHighGainVertical',
    ShortPeriodLowGainVertical = 'ShortPeriodLowGainVertical',
    ShortPeriodHighGainVertical = 'ShortPeriodHighGainVertical',
    BroadbandHighGainEastWest = 'BroadbandHighGainEastWest',
    ShortPeriodLowGainEastWest = 'ShortPeriodLowGainEastWest',
    ShortPeriodHighGainEastWest = 'ShortPeriodHighGainEastWest',
    BroadbandHighGainNorthSouth = 'BroadbandHighGainNorthSouth',
    ShortPeriodLowGainNorthSouth = 'ShortPeriodLowGainNorthSouth',
    ShortPeriodHighGainNorthSouth = 'ShortPeriodHighGainNorthSouth',
    ExtremelyShortPeriodHydrophone = 'ExtremelyShortPeriodHydrophone',
    ExtremelyShortPeriodHighGainEastWest = 'ExtremelyShortPeriodHighGainEastWest',
    ExtremelyShortPeriodHighGainNorthSouth = 'ExtremelyShortPeriodHighGainNorthSouth',
    ExtremelyShortPeriodHighGainVertical = 'ExtremelyShortPeriodHighGainVertical',
    HighBroadbandHighGainEastWest = 'HighBroadbandHighGainEastWest',
    HighBroadbandHighGainNorthSouth = 'HighBroadbandHighGainNorthSouth',
    HighBroadbandHighGainVertical = 'HighBroadbandHighGainVertical',
    MidPeriodHighGainEastWest = 'MidPeriodHighGainEastWest',
    MidPeriodHighGainNorthSouth = 'MidPeriodHighGainNorthSouth',
    MidPeriodHighGainVertical = 'MidPeriodHighGainVertical'
}

/**
 * Enumerated type of the actor (e.g. analyst, system) associated with a CreationInfo object
 */
export enum CreatorType {
    Analyst = 'Analyst',
    System = 'System'
}

/**
 * Provenance information about the results of data processing completed by the System
 * and actions completed by users of the system
 */
export interface CreationInfo {
    id: string;
    creationTime: number;
    creatorId: string;
    creatorType: CreatorType;
    creatorName: string;

    // TODO
    // The processing step associated with the processing result
    // processingStepId: string;

    // TODO
    // The software component associated with the processing result
    // softwareComponentId: string!
}

/**
 * Creation info OSD representation
 */
export interface OSDCreationInfo {
    id?: string;
    creationTime: string;
    creatorId?: string;
    creatorType?: CreatorType;
    creatorName: string;

    softwareInfo: {
        name: string;
        version: string;
    };
}

/**
 * Software info
 */
export interface SoftwareInfo {
    name: string;
    version: string;
}

/**
 * Processing Context part of data structure that filtered waveform
 * control service uses to tell who/where the request came from. In our case
 * from Interactive UI user
 */
export interface ProcessingContext {
    analystActionReference: AnalystActionReference;
    processingStepReference: ProcessingStepReference;
    storageVisibility: string;
}

/**
 * Analyst action refeerence
 */
export interface AnalystActionReference {
    processingStageIntervalId: string;
    processingActivityIntervalId: string;
    analystId: string;
}

/**
 * Processing step reference
 */
export interface ProcessingStepReference {
    processingStageIntervalId: string;
    processingSequenceIntervalId: string;
    processingStepId: string;
}

/**
 * Test data paths used when reading in data
 */
export interface TestDataPaths {
    dataHome: string;
    jsonHome: string;
    fpHome: string;
    fkHome: string;
    channelsHome: string;
    additionalDataHome: string;
}

/**
 * Represents calibration information associated with a waveform
 */
export interface ProcessingCalibration {
    factor: number;
    factorError: number;
    period: number;
    timeShift: number;
}

/**
 * Represents the configured type of data source the API Gateway provides access to - values:
 * Local - The API gateway loads data from local file storage for testing purposes
 * Service - The API gateway uses services to provide access to backend (e.g. OSD) data
 */
export enum AccessorDataSource {
    Local = 'Local',
    Service = 'Service'
}

/**
 * Enumerated list of source types used to compute distances to
 */
export enum DistanceSourceType {
    Event = 'Event',
    UserDefined = 'UserDefined'
}

/**
 * Distance value's units degrees or kilometers
 */
export enum DistanceUnits {
    degrees = 'degrees',
    km = 'km'
}

/**
 * Represents a distance measurement relative to a specified source location
 */
export interface DistanceToSource {

    // The distance in kilometers or degrees
    distance: number;

    // The units distance was calculated
    distanceUnits: DistanceUnits;

    // The source location
    sourceLocation: Location;

    // The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType;

    // Optional: the unique ID of the source object
    sourceId?: string;

    // Which station distance to the source
    stationId: string;
}

/**
 * Log level used to determine when the log statements display
 */
export enum LogLevel {
  INFO = 'INFO',
  DEBUG = 'DEBUG',
  WARNING = 'WARNING',
  ERROR = 'ERROR',
  DATA = 'DATA'
}

/**
 * Client log, object that describes the client log message
 */
export interface ClientLog {
  logLevel: LogLevel;
  message: string;
  time?: string;
}
