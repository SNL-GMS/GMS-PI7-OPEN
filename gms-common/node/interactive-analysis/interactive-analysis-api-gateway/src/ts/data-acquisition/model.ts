import { StationType } from '../common/model';

/**
 * Types of MetaData
 */
export enum MetadataType {
  RAW_STATION_DATA_FRAME = 'RAW_STATION_DATA_FRAME',
  TRANSFERRED_FILE_INVOICE = 'TRANSFERRED_FILE_INVOICE'
}

/**
 * Possible Statuses for MetaDataType
 */
export enum Status {
  RECEIVED = 'RECEIVED',
  SENT = 'SENT',
  SENT_AND_RECEIVED = 'SENT_AND_RECEIVED'
}

/**
 * Represents a Transferred File from the OSD
 */
export interface TransferredFile {
  priority: string;
  status: Status; // will filter by status==SENT
  receptionTime: string;
  transferTime: string;
  metadataType: MetadataType;
  metadata: RawStationDataFrame;
}

/**
 * Represents one of the TransferredFile metadata types
 */
export interface RawStationDataFrame {
  stationId: string;
  channelIds: string[];
  payloadStartTime: string;
  payloadEndTime: string;
}

/**
 * Represents a Reference Station
 */
export interface ReferenceStation {
  name: string;
  description: string;
  stationType: StationType;
  comment: string;
  source: InformationSource;
  latitude: number;
  longitude: number;
  elevation: number;
  actualChangeTime: string;
  systemChangeTime: string;
  aliases: ReferenceAlias[];
}

/**
 * Represents an InformationSource
 */
export interface InformationSource {
  originatingOrganization: string;
  informationTime: string;
  reference: string;
}

/**
 * Represents a ReferenceAlias
 */
export interface ReferenceAlias {
  id: string;
  name: string;
  status: StatusType;
  comment: string;
  actualChangeTime: string;
  systemChangeTime: string;
}

/**
 * Status Types
 */
export enum StatusType {
  Inactive = 'INACTIVE',
  Active = 'ACTIVE'
}
