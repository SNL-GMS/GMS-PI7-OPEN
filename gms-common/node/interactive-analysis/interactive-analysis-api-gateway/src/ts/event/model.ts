import { CreationInfo, ProcessingContext } from '../common/model';
import { FeatureMeasurementTypeName, FeatureMeasurementValue } from '../signal-detection/model';
import { ProcessingChannel } from '../station/model';
import { LocationSolutionOSD } from './model-osd';
import { PhaseType } from '../channel-segment/model-spectra';

/**
 * Model definitions for the event-related data API
 */

/**
 * The enumerated status of an event
 */
// TODO In OSD, this class DNE
export enum EventStatus {
    ReadyForRefinement = 'ReadyForRefinement',
    OpenForRefinement = 'OpenForRefinement',
    AwaitingReview = 'AwaitingReview',
    Complete = 'Complete'
}
/**
 * Represents an event marking the occurrence of some transient
 * source of energy in the ground, oceans, or atmosphere
 */
export interface Event {
    id: string;
    monitoringOrganization: string;
    preferredEventHypothesisHistory: PreferredEventHypothesis[];
    finalEventHypothesisHistory?: EventHypothesis[];
    hypotheses: EventHypothesis[];
    status: EventStatus; // Not in OSD!
    activeAnalystUserNames: string[]; // Not in OSD!
    currentEventHypothesis: PreferredEventHypothesis;
    modified: boolean;
}

/**
 * Represents a proposed explanation for an event, such that the set of
 * event hypotheses grouped by an Event represents the history of that event.
 */
export interface EventHypothesis {
    id: string;
    eventId: string;
     // TODO is this actually a list of ids?
     parentEventHypotheses: string[];
     rejected: boolean;
    // This is a list of Sets. Each set (for now) will
    // contain types Depth, Surface and Unrestrained entries
    // For the OSD it is an array since there will only be
    // one set.
    locationSolutionSets: LocationSolutionSet[];
    preferredLocationSolution: PreferredLocationSolution;
    associations: SignalDetectionEventAssociation[];
}

/**
 * 
 * Event Location definition
 */
export interface EventLocation {
    latitudeDegrees: number;
    longitudeDegrees: number;
    depthKm: number;
    time: number;
}

/**
 * The preferred hypothesis for the event at a given processing stage
 */
export interface PreferredEventHypothesis {
    processingStageId: string;
    eventHypothesis: EventHypothesis;
}

/**
 * Represents the linkage between Event Hypotheses and Signal Detection Hypotheses.
 * The rejected attribute is used to ensure that any rejected associations will not
 * be re-formed in subsequent processing stages.
 */
export interface SignalDetectionEventAssociation {
    id: string;
    signalDetectionHypothesisId: string;
    eventHypothesisId: string;
    rejected: boolean;
}

/**
 * Represents a final hypothesis for the event.
 */
export interface FinalEventHypothesis {
    eventHypothesis: EventHypothesis;
}

/**
 * Represents an estimate of the location of an event, defined as latitude, longitude, depth, and time.
 * A location solution is often determined by a location algorithm that minimizes the difference between
 * feature measurements (usually arrival time, azimuth, and slowness) and corresponding feature predictions.
 */
export interface LocationSolution {
    id: string;
    location: EventLocation;
    featurePredictions: FeaturePrediction[];
    locationRestraint: LocationRestraint;
    locationUncertainty?: LocationUncertainty;
    locationBehaviors: LocationBehavior[];
    snapshots: SignalDetectionSnapshot[];
}

/**
 * Location Solution Set
 * defines a list of location solutions for an event hypotheis
 * including a snapshot of association when solutions were created
 */
export interface LocationSolutionSet {
    id: string;
    count: number;
    locationSolutions: LocationSolution[];
}

/**
 * Snapshot of state of associations when location solution was created
 */
export interface SignalDetectionSnapshot {
    signalDetectionId: string;
    signalDetectionHypothesisId: string;
    stationId: string;
    stationName: string;
    channelName: string;
    phase: PhaseType;
    time: EventSignalDetectionAssociationValues;
    slowness: EventSignalDetectionAssociationValues;
    azimuth: EventSignalDetectionAssociationValues;
}

/**
 * Generic interface for snapshot values of a signal detection association
 */
export interface EventSignalDetectionAssociationValues {
    defining: boolean;
    observed: number;
    residual: number;
    correction: number;
}

/**
 * Represents a Feature Prediction as part of the Location Solution. This should represent a 
 * predicted location of the event.
 */
export interface FeaturePrediction {
    id: string; // not in std data
    phase: string;
    featurePredictionComponents: FeaturePredictionComponent[]; /* Java declares as a set not sure in JSON */
    extrapolated: boolean;
    predictionType: FeatureMeasurementTypeName;
    sourceLocation: EventLocation;
    receiverLocation: Location;
    channelId?: string;
    stationId?: string;
    predictedValue?: FeatureMeasurementValue;
}

/**
 * Location Restraint for event Location Solution
 */
export interface LocationRestraint {
    depthRestraintType: DepthRestraintType;
    depthRestraintKm: number;
    latitudeRestraintType: RestraintType;
    latitudeRestraintDegrees: number;
    longitudeRestraintType: RestraintType;
    longitudeRestraintDegrees: number;
    timeRestraintType: RestraintType;
    timeRestraint: string;
}

/**
 * Location Behavior for event Location Solution
 */
export interface LocationBehavior {
    residual: number;
    weight: number;
    defining: boolean;
    featurePredictionId: string;
    featureMeasurementId: string;
}

/**
 * Location Uncertainty for Location Solution
 */
export interface LocationUncertainty {
    xy: number;
    xz: number;
    xt: number;
    yy: number;
    yz: number;
    yt: number;
    zz: number;
    zt: number;
    tt: number;
    stDevOneObservation: number;
    ellipses: Ellipse [];
    ellipsoids: Ellipsoid [];
}

/**
 * Ellipse for Location Solution
 */
export interface Ellipse {
    scalingFactorType: ScalingFactorType;
    kWeight: number;
    confidenceLevel: number;
    majorAxisLength: string;
    majorAxisTrend: number;
    minorAxisLength: string;
    minorAxisTrend: number;
    depthUncertainty: number;
    timeUncertainty: string;
}
/**
 * Ellipsoid for Location Solution
 */
export interface Ellipsoid {
    scalingFactorType: ScalingFactorType;
    kWeight: number;
    confidenceLevel: number;
    majorAxisLength: number;
    majorAxisTrend: number;
    majorAxisPlunge: number;
    intermediateAxisLength: number;
    intermediateAxisTrend: number;
    intermediateAxisPlunge: number;
    minorAxisLength: number;
    minorAxisTrend: number;
    minorAxisPlunge: number;
    depthUncertainty: number;
    timeUncertainty: string;
}

/**
 * RestraintType for Location Restraint
 */
export enum RestraintType {
    UNRESTRAINED = 'UNRESTRAINED',
    FIXED = 'FIXED'
}

/**
 * DepthRestraintType for Location Restraint
 */
export enum DepthRestraintType {
    UNRESTRAINED = 'UNRESTRAINED',
    FIXED_AT_DEPTH = 'FIXED_AT_DEPTH',
    FIXED_AT_SURFACE = 'FIXED_AT_SURFACE'
}

/**
 * ScalingFactorType in  Ellipse anbd Ellipsoid
 */
export enum ScalingFactorType {
    CONFIDENCE = 'CONFIDENCE',
    COVERAGE = 'COVERAGE',
    K_WEIGHTED = 'K_WEIGHTED'
}

/**
 * Feature Prediction Component definition
 */
export interface FeaturePredictionComponent {
    value: DoubleValue;
    extrapolated: boolean;
    predictionComponentType: FeaturePredictionCorrectionType;
}

/**
 * Feature Prediction Corrections
 */
export interface FeaturePredictionCorrection {
    correctionType: FeaturePredictionCorrectionType;
    usingGlobalVelocity: boolean;
}

/**
 * Enumerated types for the Feature Prediction Corrections
 */
export enum FeaturePredictionCorrectionType {
    BASELINE_PREDICTION = 'BASELINE_PREDICTION',
    ELEVATION_CORRECTION = 'ELEVATION_CORRECTION',
    ELLIPTICITY_CORRECTION = 'ELLIPTICITY_CORRECTION'
}

/**
 * Double Value used in OSD common objects
 */
export interface DoubleValue {
    value: number;
    standardDeviation: number;
    units: Units;
}

/**
 * Units used in DoubleValue part of feature prediction
 */
export enum Units {
    DEGREES = 'DEGREES',
    SECONDS= 'SECONDS',
    SECONDS_PER_DEGREE = 'SECONDS_PER_DEGREE',
    UNITLESS = 'UNITLESS'
}

/**
 * Feature Prediction input definition for streaming call to compute FP
 */
export interface FeaturePredictionStreamingInput {
    featureMeasurementTypes: FeatureMeasurementTypeName[];
    sourceLocation: LocationSolutionOSD;
    receiverLocations: ProcessingChannel[];
    phase: string;
    model: string;
    corrections?: FeaturePredictionCorrection [];
    processingContext: ProcessingContext;
}

/**
 * Enumerated type of magnitude solution (surface wave, body wave, local, etc.)
 */
// TODO In OSD, this class DNE
export enum MagnitudeType {
    mb = 'mb',
    mbMLE = 'mbMLE',
    mbrel = 'mbrel',
    ms = 'ms',
    msMLE = 'msMLE',
    msVMAX = 'msVMAX',
    ml = 'ml'
}

/**
 * Represents an estimate of an event's magnitude based on detections from multiple stations.
 */
// TODO In OSD, this class DNE
export interface NetworkMagnitudeSolution {
    magnitudeType: MagnitudeType;
    magnitude: number;
}

/**
 * Represents a preference relationship between an event hypothesis and a location solution.
 * Creation information is included in order to capture provenance of the preference.
 */
export interface PreferredLocationSolution {
    locationSolution: LocationSolution;
    creationInfo: CreationInfo; // in OSD: DNE
}

/**
 * Encapsulates a set of event field values to apply to an event.
 */
export interface UpdateEventInput {
    creatorId: string;
    processingStageId: string;
    status: EventStatus; // TODO: field DNE in OSD
    preferredHypothesisId: string;
    activeAnalystUserNames: string[]; // TODO: field DNE in OSD
}

/**
 * Enum of Event Location Algorithms
 */
export enum EventLocationAlgorithm {
    GeigersAlgorithm = 'Geigers',
    ApacheLmAlgorithm = 'ApacheLm'
}
/**
 * Input for locate event
 */
export interface LocateEventParameter {
    pluginInfo: {
        name: string;
        version: string;
    };
    eventLocationDefinition: EventLocationDefinition;
}

/**
 * Event Location Definition base definition (ApacheLM)
 */
export interface EventLocationDefinition {
    type: string;
    maximumIterationCount: number;
    convergenceThreshold: number;
    uncertaintyProbabilityPercentile: number;
    earthModel: string;
    applyTravelTimeCorrections: boolean;
    scalingFactorType: ScalingFactorType;
    kWeight: number;
    aprioriVariance: number;
    minimumNumberOfObservations: number;
    locationRestraints: LocationRestraint[];
}

/**
 * Geigers Event Location Definition. Additional fields to ApacheLm defintion
 */
export interface GeigersEventLocationDefinition extends EventLocationDefinition {
    dampingFactorStep: number;
    deltamThreshold: number;
    depthFixedIterationCount: number;
    lambda0: number;
    lambdaX: number;
    singularValueWFactor: number;
}
/**
 * Encapsulates input used to create a new event hypothesis.
 */
//
export interface CreateEventHypothesisInput {
    associatedSignalDetectionIds: string[];
    eventLocation: EventLocation;
    creatorId: string;
    processingStageId: string;
}

/**
 * Encapsulates input used to update an existing event hypothesis.
 */
export interface UpdateEventHypothesisInput extends CreateEventHypothesisInput {
    rejected: boolean;
}
/**
 * Bundles into a touple 
 */
export interface PreferredEventHypothesisHistoryAndHypothesis {
    preferredEventHypothesisHistory: PreferredEventHypothesis[];
    currentPrefEventHypo: PreferredEventHypothesis;
}
