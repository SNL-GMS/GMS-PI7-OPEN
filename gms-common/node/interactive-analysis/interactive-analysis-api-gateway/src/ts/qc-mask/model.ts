import { TimeRange } from '../common/model';

/**
 * qc-mask categories 
 */
export enum QcMaskCategory {
    AnalystDefined = 'ANALYST_DEFINED',
    ChannelProcessing = 'CHANNEL_PROCESSING',
    DataAuthentication = 'DATA_AUTHENTICATION',
    Rejected = 'REJECTED',
    StationSOH = 'STATION_SOH',
    WaveformQuality = 'WAVEFORM_QUALITY'
}

/**
 * Qc-Mask types
 */
export enum QcMaskType {
    SensorProblem = 'SENSOR_PROBLEM',
    StationProblem = 'STATION_PROBLEM',
    Calibration = 'CALIBRATION',
    StationSecurity = 'STATION_SECURITY',
    Timing = 'TIMING',
    RepairableGap = 'REPAIRABLE_GAP',
    RepeatableAdjacentAmplitudeValue = 'REPEATED_ADJACENT_AMPLITUDE_VALUE',
    LongGap = 'LONG_GAP',
    Spike = 'SPIKE'
}

/**
 * Represents a QC Mask
 */
export interface QcMask {
    id: string;
    channelId: string;
    qcMaskVersions: QcMaskVersion[];
}

/**
 * Represents a QC Mask version entry
 */
export interface QcMaskVersion {
    version: number;
    parentQcMasks: ParentQcMask[];
    channelSegmentIds: string[];
    category: string;
    type: string;
    rationale: string;
    startTime: string;
    endTime: string;
    creationInfoId: string;
}

/**
 * Represents a QC Mask parent
 */
interface ParentQcMask {
    qcMaskId: string;
    qcMaskVersionId: number;
}

/**
 * Represents a QcMask Input
 */
export interface QcMaskInput {
    timeRange: TimeRange;
    category: QcMaskCategory;
    type: QcMaskType;
    rationale: string;
}
