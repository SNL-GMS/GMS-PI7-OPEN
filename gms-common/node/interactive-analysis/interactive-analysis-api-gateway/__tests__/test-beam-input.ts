import { ChannelSegmentProcessor } from '../src/ts/channel-segment/channel-segment-processor';
import { SignalDetection } from '../src/ts/signal-detection/model';
import { signalDetectionProcessor } from '../src/ts/signal-detection/signal-detection-processor';
import { TimeRange } from '../src/ts/common/model';
import { isOSDWaveformChannelSegment, OSDChannelSegment } from '../src/ts/channel-segment/model';
import { OSDWaveform } from '../src/ts/waveform/model';

// No need for pre-test setup yet...
// beforeEach(async () => await setupTest());
// No need for pre-test setup yet...
beforeEach(async () => await setupTest());
let timeRange: TimeRange;
let stationIds: string[];
async function setupTest() {
  timeRange = {
    startTime: 1274385600,
    endTime: 1274400000
  };
  stationIds = ['station1-1111-1111-1111-111111111111'];
  await signalDetectionProcessor.loadSignalDetections(timeRange, stationIds);
}

// ---- Method call to Channel Segment Processor ----
// Test case - Create Beam Input Parameter to send computeBeam streaming call
it('Creating new Beam Input Parameter should match snapshot', async () => {
  const sd: SignalDetection = signalDetectionProcessor.
    getSignalDetectionById('sd111111-1111-1111-1111-111111111110');
  const beamInput = await ChannelSegmentProcessor.Instance().buildBeamInput(sd);
  // Clear out the creationTime since that is current time
  if (isOSDWaveformChannelSegment(beamInput.waveforms[0])) {
    const osdWaveform = beamInput.waveforms[0] as OSDChannelSegment<OSDWaveform>;
    osdWaveform.creationInfo.creationTime = '2010-05-20T20:40:04.350Z';

    // TODO: Not sure why waveform data is coming back differently each time need
    // TODO: to figure out.
    osdWaveform.timeseries[0].values = [];
  }

  // Compare response to snapshot
  expect(beamInput).toMatchSnapshot();
});

// tslint:disable-next-line:no-empty
afterAll(async () => {});
