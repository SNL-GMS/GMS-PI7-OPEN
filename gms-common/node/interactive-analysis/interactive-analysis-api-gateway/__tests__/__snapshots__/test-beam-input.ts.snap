// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating new Beam Input Parameter should match snapshot 1`] = `
Object {
  "beamDefinition": Object {
    "beamPoint": Object {
      "elevationKm": 0.627,
      "latDegrees": -23.66513,
      "lonDegrees": 133.90526,
    },
    "coherent": true,
    "minimumWaveformsForBeam": 1,
    "nominalWaveformSampleRate": 20,
    "phaseType": "P",
    "relativePositionsByChannelId": Object {
      "channel1-1111-1111-1111-111111111110": Object {
        "eastDisplacementKm": 0,
        "northDisplacementKm": 0,
        "verticalDisplacementKm": 0,
      },
    },
    "snappedSampling": true,
    "twoDimensional": true,
    "waveformSampleRateTolerance": 0.01,
  },
  "outputChannelId": "station1-1111-1111-1111-111111111111",
  "waveforms": Array [
    Object {
      "channelId": "channel1-1111-1111-1111-111111111110",
      "creationInfo": Object {
        "creationTime": "2010-05-20T20:40:04.350Z",
        "creatorName": "Spencer",
        "softwareInfo": Object {
          "name": "Default Name",
          "version": "Default Version",
        },
      },
      "endTime": "2010-05-20T20:45:04.350Z",
      "id": "de59f588-5f41-3f82-b8f1-2d1c20d0471f",
      "name": "AS01/SHZ ACQUIRED",
      "startTime": "2010-05-20T20:40:04.350Z",
      "timeseries": Array [
        Object {
          "sampleCount": 5999,
          "sampleRate": 20,
          "startTime": "2010-05-20T20:40:04.350Z",
          "values": Array [],
        },
      ],
      "timeseriesType": "WAVEFORM",
      "type": "ACQUIRED",
    },
  ],
}
`;
